### ISSUE TITLE
promotion of 2.03.06 : develop->master

### ISSUE THREAD

**Issue by hcedwar** [kokkos/kokkos#922]
Track progress and issues for promoting develop->master, resulting in v 2.03.06

**Comment by hcedwar** [comment]
Waiting on #921 and #818

**Comment by hcedwar** [comment]
Revised OpenMP back-end **requires** environment variables be set for desired performance with applications:
```bash
export OMP_PROC_BIND=spread
export OMP_PLACES=threads
```
This must be well-advertised when the promotion completes.
For unit testing
```bash
export OMP_PROC_BIND=false
export OMP_PLACES=""
```

**Comment by hcedwar** [comment]
Need to improve warning message to be very clear about OpenMP back-end environment requirements.

**Comment by hcedwar** [comment]
Snapshot into Trilinos will require updates to match removal of `Experimental` namespace.
@ibaned has these Trilinos commits ready.

**Comment by hcedwar** [comment]
All outstanding pull requests for this promotion are done, @ibaned @ndellingwood can start tests

**Comment by ibaned** [comment]
starting tests with commit 2cd8e4e26344e1fee0cfbdb268d4e36447f43a7a

**Comment by ibaned** [comment]
Got a compile error using `test_all_sandia` on `kokkos-dev`. Assigned #926 to @ndellingwood .

**Comment by ibaned** [comment]
Restarting tests with e01dbcf3c286f0f57257b56a9e0be621394c5ace

**Comment by ibaned** [comment]
CUDA build errors. Assigned #929 to myself.

**Comment by ibaned** [comment]
New `core` unit test uses `containers` types. Assigned #930 to @ndellingwood.

**Comment by ibaned** [comment]
Latest two issues fixed, restarting with ace5c9904a7c0055029f2b33cdfb7690535e0212

**Comment by ibaned** [comment]
Most `test_all_sandia` have passed (others are still running), but Trilinos has compile failures. Sacado+Panzer issues. Really not a fan of specialized views for AD types. Working on a fix...

**Comment by ibaned** [comment]
@ndellingwood is helping with the changes to Trilinos as he is most knowledgeable regarding what is needed in Sacado

**Comment by ndellingwood** [comment]
@ibaned Pushed the fix to the kokkos-2.03.06 branch - I needed to make some corresponding changes to the `createLayout` and `computeRank` specializations in Sacado which were affected by the changes I made for issue #566

**Comment by ibaned** [comment]
Tests restarting with trilinos/Trilinos@244edc14b39916fa40ab07391f1aa4e86e523eef

**Comment by ibaned** [comment]
All `test_all_sandia` tests have passed on `kokkos-dev` for ace5c9904a7c0055029f2b33cdfb7690535e0212
```
clang-3.6.1-Pthread-hwloc-release build_time=107 run_time=89
clang-3.6.1-Pthread-release build_time=90 run_time=130
clang-3.6.1-Pthread_Serial-hwloc-release build_time=108 run_time=120
clang-3.6.1-Pthread_Serial-release build_time=112 run_time=148
clang-3.6.1-Serial-hwloc-release build_time=95 run_time=105
clang-3.6.1-Serial-release build_time=150 run_time=90
clang-3.7.1-Pthread-hwloc-release build_time=143 run_time=62
clang-3.7.1-Pthread-release build_time=124 run_time=155
clang-3.7.1-Pthread_Serial-hwloc-release build_time=119 run_time=115
clang-3.7.1-Pthread_Serial-release build_time=113 run_time=188
clang-3.7.1-Serial-hwloc-release build_time=96 run_time=104
clang-3.7.1-Serial-release build_time=116 run_time=105
clang-3.8.1-Pthread-hwloc-release build_time=128 run_time=78
clang-3.8.1-Pthread-release build_time=159 run_time=129
clang-3.8.1-Pthread_Serial-hwloc-release build_time=134 run_time=122
clang-3.8.1-Pthread_Serial-release build_time=156 run_time=181
clang-3.8.1-Serial-hwloc-release build_time=157 run_time=94
clang-3.8.1-Serial-release build_time=105 run_time=128
clang-3.9.0-Pthread-hwloc-release build_time=130 run_time=60
clang-3.9.0-Pthread-release build_time=179 run_time=129
clang-3.9.0-Pthread_Serial-hwloc-release build_time=146 run_time=94
clang-3.9.0-Pthread_Serial-release build_time=173 run_time=158
clang-3.9.0-Serial-hwloc-release build_time=146 run_time=109
clang-3.9.0-Serial-release build_time=112 run_time=121
cuda-7.0.28-Cuda_OpenMP-release build_time=421 run_time=594
cuda-7.0.28-Cuda_Pthread-release build_time=418 run_time=569
cuda-7.0.28-Cuda_Serial-release build_time=388 run_time=585
cuda-7.5.18-Cuda_OpenMP-release build_time=391 run_time=579
cuda-7.5.18-Cuda_Pthread-release build_time=385 run_time=576
cuda-7.5.18-Cuda_Serial-release build_time=371 run_time=577
cuda-8.0.44-Cuda_OpenMP-release build_time=363 run_time=575
cuda-8.0.44-Cuda_Pthread-release build_time=363 run_time=572
cuda-8.0.44-Cuda_Serial-release build_time=345 run_time=573
gcc-4.7.2-OpenMP-hwloc-release build_time=160 run_time=81
gcc-4.7.2-OpenMP-release build_time=160 run_time=94
gcc-4.7.2-OpenMP_Serial-hwloc-release build_time=157 run_time=130
gcc-4.7.2-OpenMP_Serial-release build_time=171 run_time=111
gcc-4.7.2-Pthread-hwloc-release build_time=122 run_time=55
gcc-4.7.2-Pthread-release build_time=157 run_time=77
gcc-4.7.2-Pthread_Serial-hwloc-release build_time=142 run_time=157
gcc-4.7.2-Pthread_Serial-release build_time=152 run_time=121
gcc-4.7.2-Serial-hwloc-release build_time=100 run_time=96
gcc-4.7.2-Serial-release build_time=104 run_time=93
gcc-4.8.4-OpenMP-hwloc-release build_time=123 run_time=135
gcc-4.8.4-OpenMP-release build_time=122 run_time=132
gcc-4.8.4-OpenMP_Serial-hwloc-release build_time=135 run_time=158
gcc-4.8.4-OpenMP_Serial-release build_time=135 run_time=129
gcc-4.8.4-Pthread-hwloc-release build_time=121 run_time=62
gcc-4.8.4-Pthread-release build_time=117 run_time=128
gcc-4.8.4-Pthread_Serial-hwloc-release build_time=124 run_time=90
gcc-4.8.4-Pthread_Serial-release build_time=115 run_time=178
gcc-4.8.4-Serial-hwloc-release build_time=115 run_time=97
gcc-4.8.4-Serial-release build_time=101 run_time=126
intel-14.0.4-OpenMP-hwloc-release build_time=341 run_time=279
intel-14.0.4-OpenMP-release build_time=330 run_time=240
intel-14.0.4-OpenMP_Serial-hwloc-release build_time=328 run_time=244
intel-14.0.4-OpenMP_Serial-release build_time=355 run_time=224
intel-14.0.4-Pthread-hwloc-release build_time=336 run_time=193
intel-14.0.4-Pthread-release build_time=305 run_time=253
intel-14.0.4-Pthread_Serial-hwloc-release build_time=315 run_time=199
intel-14.0.4-Pthread_Serial-release build_time=314 run_time=242
intel-14.0.4-Serial-hwloc-release build_time=300 run_time=223
intel-14.0.4-Serial-release build_time=299 run_time=230
intel-15.0.2-OpenMP-hwloc-release build_time=342 run_time=193
intel-15.0.2-OpenMP-release build_time=338 run_time=193
intel-15.0.2-OpenMP_Serial-hwloc-release build_time=381 run_time=238
intel-15.0.2-OpenMP_Serial-release build_time=381 run_time=236
intel-15.0.2-Pthread-hwloc-release build_time=281 run_time=178
intel-15.0.2-Pthread-release build_time=297 run_time=195
intel-15.0.2-Pthread_Serial-hwloc-release build_time=322 run_time=229
intel-15.0.2-Pthread_Serial-release build_time=286 run_time=254
intel-15.0.2-Serial-hwloc-release build_time=311 run_time=180
intel-15.0.2-Serial-release build_time=247 run_time=204
intel-16.0.1-OpenMP-hwloc-release build_time=412 run_time=249
intel-16.0.1-OpenMP-release build_time=408 run_time=239
intel-16.0.1-OpenMP_Serial-hwloc-release build_time=399 run_time=240
intel-16.0.1-OpenMP_Serial-release build_time=421 run_time=243
intel-16.0.1-Pthread-hwloc-release build_time=325 run_time=195
intel-16.0.1-Pthread-release build_time=339 run_time=240
intel-16.0.1-Pthread_Serial-hwloc-release build_time=307 run_time=239
intel-16.0.1-Pthread_Serial-release build_time=326 run_time=228
intel-16.0.1-Serial-hwloc-release build_time=297 run_time=202
intel-16.0.1-Serial-release build_time=287 run_time=216
```

**Comment by ibaned** [comment]
Working through lots more Sacado issues. See #934.

**Comment by hcedwar** [comment]
Blocker #935

**Comment by ibaned** [comment]
restarting integration tests with 26c639d and trilinos/Trilinos@fa73121

**Comment by ibaned** [comment]
One more build error, in the Sacado subdirectory of Stokhos (why does such a thing exist).

Good news is, the Trilinos integration tests are starting to pass! The CUDA build on `white` is totally clean for both `updated` and `pristine`. Here is a diff of failing tests, with `pristine` on the left (`<`) and `updated` on the right (`>`):
```
Diff Failed Tests
47d46
< Anasazi_IRTRThyra_test_1_MPI_4 (Failed)
61a61
> MueLu_Driver_TogglePFactory_semi_sa_line_easy_Tpetra_MPI_4 (Failed)
70,73d69
< PanzerDiscFE_tFieldPattern_MPI_4 (Failed)
< PanzerDofMgr_tFieldPattern_MPI_4 (Failed)
< ROL_adapters_thyra_test_vector_ScaledThyraVector_MPI_4 (Failed)
< ROL_adapters_thyra_test_vector_ThyraInterface_MPI_4 (Failed)
75,99d70
< ROL_test_function_BarrierFunctionCheck_MPI_1 (Failed)
< ROL_test_function_BlockOperatorCheck_MPI_1 (Failed)
< ROL_test_function_CompositeEqualityConstraintCheck_MPI_1 (Failed)
< ROL_test_function_CompositeObjectiveCheck_MPI_1 (Failed)
< ROL_test_function_CompositeObjectiveSimOptCheck_MPI_1 (Failed)
< ROL_test_function_DerivativeCheck_MPI_1 (Failed)
< ROL_test_function_EqualityConstraintPartionedCheck_MPI_1 (Failed)
< ROL_test_function_HouseholderReflectorCheck_MPI_1 (Failed)
< ROL_test_function_NonlinearLeastSquaresCheck_MPI_1 (Failed)
< ROL_test_function_operator_StdLinearOperator_MPI_1 (Failed)
< ROL_test_function_operator_StdTridiagonalOperator_MPI_1 (Failed)
< ROL_test_function_SimOptSolveCheck_MPI_1 (Failed)
< ROL_test_function_StdObjectiveCheck_MPI_1 (Failed)
< ROL_test_step_BoxConstrained_LineSearch_MPI_1 (Failed)
< ROL_test_step_BoxConstrained_PrimalDualActiveSet_MPI_1 (Failed)
< ROL_test_step_BoxConstrained_TrustRegion_MPI_1 (Failed)
< ROL_test_step_LineSearch_MPI_1 (Failed)
< ROL_test_step_TrustRegion_MPI_1 (Failed)
< ROL_test_vector_CArrayVectorInterface_MPI_1 (Failed)
< ROL_test_vector_MultiVectorInterface_MPI_1 (Failed)
< ROL_test_vector_PartitionedVectorInterface_MPI_1 (Failed)
< ROL_test_vector_ProfiledVectorInterface_MPI_1 (Failed)
< ROL_test_vector_RieszVectorInterface_MPI_1 (Failed)
< ROL_test_vector_ScaledStdVectorInterface_MPI_1 (Failed)
< ROL_test_vector_StdVectorInterface_MPI_1 (Failed)
```
Note that a bunch of tests stopped failing after the update, so thats good. The only test that started failing is `MueLu_Driver_TogglePFactory_semi_sa_line_easy_Tpetra_MPI_4`.

**Comment by ibaned** [comment]
Note: Trilinos `updated` is trilinos/Trilinos@3a0523c8f755ec74e434ca419cc1ea388b0b58c8, Trilinos `pristine` is trilinos/Trilinos@0381f35c3fb278f8d77d5b91fe9233e1b11e928a, and Kokkos is 63c0ad763a05a90e7210b7c035624a9bc209997f

**Comment by crtrott** [comment]
Just as a note: I did some miniApp performance testing, and things look good (i.e. no performance regression). One thing I noticed though is a roughly 5% increase in binary size 👎

**Comment by crtrott** [comment]
Also assuming this gets down by Wednesday the new release will be 2.03.13, after Wednesday It will be 14.

**Comment by nmhamster** [comment]
Can you please make sure you include #954?

**Comment by ndellingwood** [comment]
Merged #971; as discussed the performance tests of concern on Power will be tracked in a separate issue;
@hcedwar @crtrott pending inclusion #954 mentioned in above comment, integration testing should be ready to resume.

**Comment by ndellingwood** [comment]
Promotion complete, closing.
